sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2349,7 +2350,31 @@ class Analyzer(
   }
 
   /**
-   * Replace [[TimeZoneAwareExpression]] without timezone id by its copy with session local
+   * Replace unresolved expressions in grouping keys with resolved ones in SELECT clauses.
+   */
+  object ResolveAggAliasInGroupBy extends Rule[LogicalPlan] {
+
+    override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperators {
+      case agg @ Aggregate(groups, aggs, child)
+          if conf.groupByAliases && child.resolved && groups.exists(!_.resolved) =>
+        agg.copy(groupingExpressions = groups.map {
+          case u: UnresolvedAttribute =>
+            val resolvedAgg = aggs.find(ne => resolver(ne.name, u.name))
+            // Check if no aggregate function exists in GROUP BY
```

Comment:
Since `ResolveOrdinalInOrderByAndGroupBy` has [the possibly similar error handling](https://github.com/maropu/spark/blob/21a5c85811fbdd7f10ac5d9f57552c5865c4be5e/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L993), we need to move both code into `CheckAnalysis` if we do.